### PR TITLE
Provide samples of seeding random engines in C++.

### PR DESCRIPTION
**Pull request by socantre**
Show three ways of seeding standard random engines in C++, progressing from the simplest to more advanced methods that provide more possible results.

### CODE REVIEW

**Comment by sftrabbit** [comment]
Thanks! Will review later today.

**Comment by sftrabbit** [comment]
Thanks very much for writing this.

I think it might be a little too complex for a single sample. A sample should really capture a single generic feature or idiom from modern C++. This is more like a collection of examples of how to seed RNGs.

I think a sample showing and explaining `std::seed_seq` would be quite useful. A sample showing how to seed with time might have potential too, although I'm not sure if seeding with time should be a recommended idiom (especially when we have `std::random_device`).

Would you be interested in cutting this down to something simpler?

**Comment by socantre** [comment]
Indeed, my concern with showing just a single example would be that it would gives the impression of a single 'right' way that's sufficient for all uses. I sort of wanted to give the impression of a spectrum.

I suppose if I were to give one example it would be using `seed_seq` with several usages of random_device.

I wouldn't recommend seeding with a single time value because it's not a great idea and because in C++ it's pretty verbose. Including time as one of many input's to a `seed_seq` is okay, if not really necessary; the two cases that matter with seed generation is; completely reproducible between different machines, times, etc. and unpredictable, for which random_device should be sufficient.

I'll think about how to change this sample. Do you want to limit samples to a line limit or do you really want just a single example of seeding?

**Comment by sftrabbit** [comment]
There's no real line limit. The "rule of five" sample is quite long, but simply because it has to be. It's still just explaining one thing (even though there will be various ways of implementing the rule of five, depending on the class). I just want the samples to convey how to perform some particular common task. Providing multiple examples of how to use a particular class/feature is more what I'd expect from the examples on cppreference.com.

**Comment by socantre** [comment]
Alright, I've simplified this sample.

**Comment by sftrabbit** [comment]
Thanks for the sample. Definitely much more appropriate now! I've reviewed and edited it ([final version here](https://github.com/sftrabbit/CppSamples-Samples/blob/master/1-common-tasks/random/high-entropy-seed.cpp)). I just reworded it a bit to try to capture what exactly `seed_seq` does and make it easy to understand. Let me know if I've screwed anything up - I'm certainly not a pseudo-random number generation expert!

**Comment by socantre** [comment]
I posted some comments on the commit:

https://github.com/sftrabbit/CppSamples-Samples/commit/36d3eaa24910ae85ea6e5001897ff0259ab18c9b

**Comment by sftrabbit** [comment]
Also posted some comments. Will work on a revised version.

**Comment by sftrabbit** [comment]
Let me know your thoughts on [the revised description](https://github.com/sftrabbit/CppSamples-Samples/blob/6810784b41d19e9940ba38064afb06e73150dbff/1-common-tasks/random/high-entropy-seed.cpp). This is certainly one of the more difficult ones to get right. It doesn't help that there's a lack of information out there about using seed_seq!

**Comment by socantre** [comment]
I do like the new description better.

**Comment by socantre** [comment]
Actually, the title shouldn't be 'unpredictable random numbers' since the predictability of the numbers themselves depends on the engine. E.g., you can completely predict the results from `std::mt19937` after seeing 624 results from it. We're just _seeding_ unpredictably: a random number engine has some number of possible output streams and the seed in effect picks one of them. So seeding using 32 bits means you pick one of 2<sup>32</sup> possibilities, which is a small enough subset of Mersenne Twister's 2<sup>19968</sup> total possible states that it's much easier to figure out what stream of numbers will be generated. (Actually, mt19937 only has 2<sup>31</sup> streams, but the period is so long that picking different places in the streams is usually just about as good as having more streams, so I focus on the total state size rather than stream count.)

Also, six calls to `r()` isn't exactly arbitrary, but if you want the the /u/ProfONiell approved number of calls, [it's 8](http://www.reddit.com/r/cpp/comments/32u4m7/the_behavior_of_rng_seeding_in_c_may_surprise/cqexc14).

**Comment by sftrabbit** [comment]
The issue I have is that putting "unpredictable" in the title at all will suggest to the uninformed reader that they'll get "more random" random numbers in some way, which isn't really true and is especially dangerous when they're being shown how to do this with `std::mt19937`. And then I feel like if we explain "you get a greater variety of initial states, but that doesn't necessarily mean that your random numbers will be unpredictable, cryptographically secure, or of high statistical quality", the reader will just think "well why would I want this then?" I think most readers of the site will just have some concept of "bad random" and "good random" and would expect this sample to give them "good random". Unfortunately, we can't really get "good random" without leaving the standard library, so how can we make this sample clear and useful?

**Comment by socantre** [comment]
I think it's a mistake to put the onus for teaching about cryptography or statistical quality on this example. Indeed, if you want to cover those, they'd probably fit better in one of the samples that covers using the engine: someone might figure they can use the 'roll a die' sample to create a diceware password, for example.

The goal is simply to displace `std::mt19937 eng{std::random_device{}()};` as the default advice on seeding when users figure out that they need seeding, not to teach them why they need seeding, or about the subtle issues that differentiate this method of seeding from that method. The goal is simply that when a new user realizes 'Hey, I'm getting the same results every time I run the program, how do I fix that?' someone who finds this page will get a reasonably easy, standard, portable method that fixes the issue they see as well as 'secretly' steering them away from solutions that could have a couple other problems.

My initial submission did that by showing the different methods and trying to explain why one might choose one over another. The simplification depended on abandoning the idea of comparison between different methods or explanation of the issues, and simply picked the most reasonable default method.

Basically, the goal with the simplified example is simply that users never see or consider `std::mt19937 eng{std::random_device{}()};`.

---

So I also would not have a paragraph talking about getting 'a greater variety of inital states', or anything like that. Again, this sample isn't meant to be seen in contrast to other methods that provide greater or lesser amounts of whatever characteristic. The word 'greater' shouldn't be used. If it's meant to be compared with anything, it's with completely reproducible engines, such as default initialized engines.

---

On 'bad random' vs. 'good random', again, that's really outside the scope of this sample. It might reasonably be addressed by another sample, although it seems to me like a topic more complicated than is suitable for this format.
